packages/devtools_app/lib/src/debugger/flutter/console.dart

```diff
@@ -30,7 +33,102 @@ class Console extends StatefulWidget {
 }
 
 class _ConsoleState extends State<Console> {
-  final scrollController = ScrollController();
+  @override
+  Widget build(BuildContext context) {
+    return ValueListenableBuilder<List<String>>(
+      valueListenable: widget.controller.stdio,
+      builder: (context, lines, _) {
+        // Extract this to a reusable ConsoleWithControls StatelessWidget?
+        return Stack(
+          children: [
+            ConsoleOutput(lines: lines),
+            Container(
+              alignment: Alignment.bottomRight,
+              child: ConsoleControls(
+                disabled: lines.isEmpty,
+                onCopyPressed: () {
+                  Clipboard.setData(ClipboardData(
+                    text: lines.join('\n'),
+                  )).then((_) {
+                    Notifications.of(context)?.push(
+                      'Copied!',
+                    );
+                  });
+                },
+                onDeletePressed: widget.controller.clearStdio,
+              ),
+            ),
+          ],
+        );
+      },
+    );
+  }
+}
+
+/// Optionally renders a ButtonBar with Console Controls,
+/// Copy and Clear.
+/// The callbacks for those buttons are passed from the outside.
+class ConsoleControls extends StatelessWidget {
+  const ConsoleControls({
+    this.disabled,
+    this.onCopyPressed,
+    this.onDeletePressed,
+  }) : super();
+
+  final Function onCopyPressed;
+  final Function onDeletePressed;
+  final bool disabled;
+
+  @override
+  Widget build(BuildContext context) {
+    return disabled
+        ? Container()
+        : ButtonBar(
+            buttonPadding: EdgeInsets.zero,
+            alignment: MainAxisAlignment.end,
+            children: [
+              IconButton(
+                icon: const Icon(Icons.content_copy),
+                onPressed: onCopyPressed,
+                tooltip: 'Copy to clipboard.',
+              ),
+              IconButton(
+                icon: const Icon(Icons.block),
+                onPressed: onDeletePressed,
+                tooltip: 'Clear console output.',
+              ),
+            ],
+          );
+  }
+}
+
+/// Renders a widget with the output of the console.
+/// This is a ListView of text lines, with a monospace font and a border.
+class ConsoleOutput extends StatefulWidget {
+  const ConsoleOutput({
+    Key key,
+    this.lines,
+  }) : super(key: key);
+
+  final List<String> lines;
+
+  @override
+  _ConsoleOutputState createState() => _ConsoleOutputState();
+}
+
+class _ConsoleOutputState extends State<ConsoleOutput> {
+  // The scroll controller must survive ConsoleOutput re-renders
```

Comment:
Thanks! This is good cleanup.

Reply:
Thanks! I want to make this console more reusable, once I get the TODO above done, I might be able to reuse this in the logging panel (so we get "Copy" there for free)